apps/dashr-oil-and-gas-ternary/app.R

```diff
@@ -0,0 +1,761 @@
+  return(htmlP(
+    className='graph-title',
+    children=title))}
+
+generate_production_plot = function(processed_data){
+  layout = list(
+    xaxis=list(
+      title='Year'
+    ),
+    yaxis=list(
+      title='GAS Production(mcf)'
+    )
+  )
+  
+  data = list()
+  
+  welllist = lapply(1:length(processed_data[['well_id']]), 
+                    function(x){list('well_id' = unlist(processed_data[['well_id']][x]), 
+                                     'formation' = unlist(processed_data[['formation']][x]))})
+  
+  for (well_id in welllist) {
+    well_prod = df_prod[df_prod$RecordNumber %in% well_id$well_id,]
+    new_trace = list(
+      x=well_prod$Year,
+      y=well_prod$VolumeMCF,
+      name=c(well_id$well_id),
+      mode='lines+markers',
+      hoverinfo='x+y+name',
+      marker=list(symbol="hexagram-open", 
+                  color=colormap[[as.character(well_id$formation)]],
+                  line=list("width" = "0.5")),
+      line=list(
+        shape='spline'
+      ),
+      showlegend=TRUE
+    )
+    data = list.append(data,new_trace)
+  }
+  return(list('data' = data, 'layout' = layout))
+  
+}
+
+generate_well_map = function(dff,selected_data,style){
+  
+  formations = as.list(unique(dff$fm_name))
+  
+  data = list()
+  
+  for (formation in unlist(formations)) {
+    selected_index = NULL
+    if(formation %in% labels(selected_data)){
+      selected_index = selected_data[[formation]]
+    }
+    
+    text_list = list(
+      paste('Well ID', dff[dff$fm_name == formation,'RecordNumber']))
+    
+    op_list = as.list(dff[dff$fm_name == formation,'op'])
+    
+    text_list = lapply(1:length(text_list[[1]]), function(x){paste(op_list[[x]], "<br>", text_list[[1]][x])})
+    
+    
+    new_trace = list(
+      list(
+        type="scattermapbox",
+        lat = dff[dff$fm_name == formation,'nlat83'],
+        lon = dff[dff$fm_name == formation,'wlon83'],
+        mode = 'markers',
+        marker = list('color' = colormap[[formation]], 'size' = 9),
+        text= unlist(text_list),
+        name = c(unlist(formation)),
+        selectedpoints = selected_index,
+        customdata = dff[dff$fm_name == formation,'RecordNumber']
+      )
+    )
+    data = append(data, new_trace)
+  }
+  
+  layout = list(
+    clickmode="event+select",
+    dragmode="lasso",
+    showlegend=TRUE,
+    autosize=TRUE,
+    hovermode="closest",
+    margin= list(l=0, r=0, t=0, b=0),
+    mapbox = list(
+      accesstoken=mapbox_access_token,
+      bearing=0,
+      center= list(lat=37.497562, lon=-82.755728),
+      pitch=0,
+      zoom=8,
+      style=style
+    ),
+    legend=list(
+      bgcolor="#1f2c56",
+      orientation="h",
+      font=list(color="white"),
+      x=0,
+      y=0,
+      yanchor="bottom"
+    )
+  )
+  
+  return(list(data = data, layout = layout))
+  
+}
+
+
+
+generate_ternary_map = function(dff, selected_data, contour_visible, marker_visible){
+  contour_traces = list()
```

Comment:
I'm probably in an extreme minority here, but I actually consider defining empty lists prior to assigning into their elements (e.g. while looping) to be a bit of an anti-pattern in R-- there are definitely contexts where it makes sense, but often you can either use `lapply` to generate a list of the appropriate dimensions, or manually define the list structure up front and then "fill in" as you go. 

I wouldn't worry about this for now, but I think it's good practice to keep in mind for the future.